File: multimodal-generation/video-generation/python/02_image_to_video_generation.py
#!/usr/bin/env python3

import boto3
import logging
import json
import random
import base64
import amazon_video_util

# Configure logging
logging.basicConfig(level=logging.INFO, format='[%(levelname)s] %(message)s')
logger = logging.getLogger(__name__)

# Constants
DEFAULT_MODEL_ID = "amazon.nova-reel-v1:0"

def setup_aws_session(region="us-east-1"):
    """Set up AWS session with default region"""
    boto3.setup_default_session(region_name=region)
    logger.info("AWS SDK session defaults have been set.")

def generate_video(s3_destination_bucket, input_image_path, video_prompt, model_id=DEFAULT_MODEL_ID):
    """
    Generate a video from an input image and prompt.
    
    Args:
        s3_destination_bucket (str): The S3 bucket where the video will be stored
        input_image_path (str): Path to the input image (must be 1280 x 720)
        video_prompt (str): Text prompt describing the desired video
    """
    # Set up the S3 client
    s3_client = boto3.client("s3")

    # Create the S3 bucket
    s3_client.create_bucket(Bucket=s3_destination_bucket)

    # Create the Bedrock Runtime client
    bedrock_runtime = boto3.client("bedrock-runtime")

    # Load the input image as a Base64 string
    with open(input_image_path, "rb") as f:
        input_image_bytes = f.read()
        input_image_base64 = base64.b64encode(input_image_bytes).decode("utf-8")

    model_input = {
        "taskType": "TEXT_VIDEO",
        "textToVideoParams": {
            "text": video_prompt,
            "images": [
                {
                    "format": "png",  # May be "png" or "jpeg"
                    "source": {
                        "bytes": input_image_base64
                    }
                }
            ]
        },
        "videoGenerationConfig": {
            "durationSeconds": 6,  # 6 is the only supported value currently
            "fps": 24,  # 24 is the only supported value currently
            "dimension": "1280x720",  # "1280x720" is the only supported value currently
            "seed": random.randint(
                0, 2147483648
            ),  # A random seed guarantees we'll get a different result each time this code runs
        },
    }

    try:
        # Start the asynchronous video generation job
        invocation = bedrock_runtime.start_async_invoke(
            modelId=model_id,
            modelInput=model_input,
            outputDataConfig={"s3OutputDataConfig": {"s3Uri": f"s3://{s3_destination_bucket}"}},
        )

        # Store the invocation ARN
        invocation_arn = invocation["invocationArn"]

        # Pretty print the response JSON
        logger.info("\nResponse:")
        logger.info(json.dumps(invocation, indent=2, default=str))

        # Save the invocation details for later reference
        amazon_video_util.save_invocation_info(invocation, model_input)

        return invocation_arn

    except Exception as e:
        logger.error(e)
        return None

def main():
    # Initialize the AWS session
    setup_aws_session()

    # Configuration
    S3_BUCKET = "nova-videos"  # Change this to your unique bucket name
    VIDEO_PROMPT = "Closeup of a large seashell in the sand, gentle waves flow around the shell. Camera zoom in."
    INPUT_IMAGE_PATH = "../images/sample-frame-1.png"  # Must be 1280 x 720
    MODEL_ID = "amazon.nova-reel-v1:0"

    # Generate the video
    invocation_arn = generate_video(S3_BUCKET, INPUT_IMAGE_PATH, VIDEO_PROMPT, MODEL_ID)
    if not invocation_arn:
        logger.error("Failed to start video generation")
        exit(1)
    
    # Monitor and download the video
    logger.info("\nMonitoring job and waiting for completion...")
    local_file_path = amazon_video_util.monitor_and_download_video(invocation_arn, "output")

    if local_file_path:
        logger.info(f"\nVideo successfully generated and downloaded to: {local_file_path}")
    else:
        logger.info("\nFailed to generate or download video")

if __name__ == "__main__":
    main()

File: multimodal-generation/video-generation/javascript/02_image_to_video_generation.js
#!/usr/bin/env node

import {
    BedrockRuntimeClient,
    StartAsyncInvokeCommand,
} from '@aws-sdk/client-bedrock-runtime';
import { S3Client, CreateBucketCommand } from '@aws-sdk/client-s3';
import winston from 'winston';
import { monitorAndDownloadVideo, saveInvocationInfo, imageToBase64 } from './amazonVideoUtil.js';

// Configure logging
const logger = winston.createLogger({
    level: 'info',
    format: winston.format.simple(),
    transports: [
        new winston.transports.Console()
    ]
});

// Constants
const DEFAULT_MODEL_ID = "amazon.nova-reel-v1:0";

// Initialize AWS clients
let bedrockRuntime;
let s3Client;

/**
 * Initializes AWS Bedrock Runtime and S3 clients.
 * @param {string} [region="us-east-1"] - The AWS region to initialize clients for.
 */
function setupAwsSession(region = "us-east-1") {
    bedrockRuntime = new BedrockRuntimeClient({ region });
    s3Client = new S3Client({ region });
    logger.info("AWS SDK clients have been initialized.");
}

/**
 * Generates a video from an image using the specified prompt and model.
 * @param {string} s3DestinationBucket - The S3 bucket to store the generated video.
 * @param {string} videoPrompt - The prompt for video generation.
 * @param {string} inputImageBase64 - The base64 encoded image to use as input.
 * @param {string} [modelId=DEFAULT_MODEL_ID] - The model ID to use for video generation.
 * @returns {Promise<string|null>} - The ARN of the invocation if successful, otherwise null.
 */
async function generateVideoFromImage(s3DestinationBucket, videoPrompt, inputImageBase64, modelId = DEFAULT_MODEL_ID) {
    try {
        // Create the S3 bucket
        await s3Client.send(new CreateBucketCommand({
            Bucket: s3DestinationBucket
        }));

        const modelInput = {
            taskType: "TEXT_VIDEO",
            textToVideoParams: {
                text: videoPrompt,
                images: [
                    {
                        format: "png",
                        source: {
                            bytes: inputImageBase64
                        }
                    }
                ]
            },
            videoGenerationConfig: {
                durationSeconds: 6,
                fps: 24,
                dimension: "1280x720",
                seed: Math.floor(Math.random() * 2147483648),
            },
        };

        console.log(modelInput)

        const command = new StartAsyncInvokeCommand({
            modelId: modelId,
            contentType: "application/json",
            accept: "application/json",
            modelInput: modelInput,
            outputDataConfig: {
                s3OutputDataConfig: {
                    s3Uri: `s3://${s3DestinationBucket}`
                }
            }
        });

        const invocation = await bedrockRuntime.send(command);
        const invocationArn = invocation.invocationArn;

        logger.info("\nResponse:");
        logger.info(JSON.stringify(invocation, null, 2));

        await saveInvocationInfo(invocation, modelInput);

        return invocationArn;

    } catch (error) {
        logger.error(error);
        return null;
    }
}

/**
 * The main function to orchestrate video generation, status checking, and downloading.
 * @returns {Promise<void>} - A promise that resolves when the main function completes.
 */
async function main() {
    // Initialize the AWS session
    setupAwsSession();

    // Configuration
    const S3_BUCKET = "nova-videos"; // Change this to your unique bucket name
    const VIDEO_PROMPT = "Closeup of a large seashell in the sand, gentle waves flow around the shell. Camera zoom in.";
    const MODEL_ID = "amazon.nova-reel-v1:0";

    const inputImageBase64 = await imageToBase64('../images/sample-frame-1.png')

    // Generate video
    const invocationArn = await generateVideoFromImage(S3_BUCKET, VIDEO_PROMPT, inputImageBase64, MODEL_ID);
    if (!invocationArn) {
        logger.error("Failed to start video generation");
        process.exit(1);
    }

    // Monitor and download the video
    logger.info("\nMonitoring job and waiting for completion...");
    const localFilePath = await monitorAndDownloadVideo(invocationArn, "output");

    if (localFilePath) {
        logger.info(`\nVideo successfully generated and downloaded to: ${localFilePath}`);
    } else {
        logger.info("\nFailed to generate or download video");
    }
}

// Self-executing async function
(async () => {
    try {
        await main();
    } catch (err) {
        console.error('Fatal error:', err.message);
        process.exit(1);
    }
})();